validation-test/stdlib/Zip.swift

```diff
@@ -0,0 +1,196 @@
+  }
+
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+    
+  for test in zipTests {
+    let left = MinimalCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
+
+    checkCollection(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, 
+      result, sameValue: compareElements)      
+  }
+}
+
+ZipTests.test("RandomAccessCollections") {
+  typealias Element = (OpaqueValue<Int>, OpaqueValue<Int32>)
+  typealias Sequence = Zip2Sequence<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+  typealias Collection = Zip2Collection<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+
+  func compareElements(_ lhs: Element, rhs: Element) -> Bool {
+    return lhs.0.value == rhs.0.value && lhs.1.value == rhs.1.value
+  }
+      
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+
+  for test in zipTests {
+    let left = MinimalRandomAccessCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalRandomAccessCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
+    checkRandomAccessCollection(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, 
+      result, sameValue: compareElements)
+  }
+}
+
+ZipTests.test("Collection.isEmpty") {
+  expectTrue(zip(0..<0,0..<0).isEmpty)
+  expectTrue(zip(0..<1,0..<0).isEmpty)
+  expectTrue(zip(0..<0,0..<1).isEmpty)
+  expectFalse(zip(0..<1,0..<1).isEmpty)
+  expectFalse(zip(0..<2,0..<3).isEmpty)
+  expectFalse(zip(0..<3,0..<2).isEmpty)
+}
+
+ZipTests.test("Collection.count") {
+  expectEqual(zip(0..<0,0..<0).underestimatedCount, 0)
+  expectEqual(zip(0..<1,0..<0).underestimatedCount, 0)
+  expectEqual(zip(0..<0,0..<1).underestimatedCount, 0)
+  expectEqual(zip(0..<2,0..<1).underestimatedCount, 1)
+  expectEqual(zip(0..<1,0..<2).underestimatedCount, 1)
+  expectEqual(zip(0..<2,0..<2).underestimatedCount, 2)
+  expectEqual(zip(0..<0,0..<0).count, 0)
+  expectEqual(zip(0..<1,0..<0).count, 0)
+  expectEqual(zip(0..<0,0..<1).count, 0)
+  expectEqual(zip(0..<2,0..<1).count, 1)
+  expectEqual(zip(0..<1,0..<2).count, 1)
+  expectEqual(zip(0..<2,0..<2).count, 2)
+}
+
+ZipTests.test("Equatable") {
+  let s = 
```

Comment:
Does his compile?!

Reply:
Huh. Yeah, with a warning that lit doesn't surface. Good spot...

Reply:
Ah.. Right. You're producing an `s: Unit`.